core/explorer/extract/cli.py

```diff
@@ -0,0 +1,366 @@
+                    logger.warning("Wrong derivation: %s", json.dumps(found_derivation))
+                    continue
+                if output_path not in queued_output_paths:
+                    queue.put(attribute_path)
+                    queued_output_paths.add(output_path)
+
+
+def process_attribute_path(
+    pipe_out: IO[str],
+    pipe_out_lock: Lock,
+    queue: multiprocessing.Queue,
+    queued_output_paths: set[str],
+    visited_output_paths: set[str],
+    finder_env: dict,
+    offline: bool,
+    attribute_path: str,
+    logger: logging.Logger,
+):
+    """
+    Describe a derivation, write results to output pipe and add potential derivations
+    to process to queue
+    """
+
+    # evaluate value at the attribute path using Nix
+    env = copy.deepcopy(finder_env)
+    env["TARGET_ATTRIBUTE_PATH"] = attribute_path
+    description_process = subprocess.run(
+        args=[
+            arg
+            for arg in [
+                "nix",
+                "eval",
+                "--offline" if offline else None,
+                "--extra-experimental-features",
+                "nix-command flakes",
+                "--json",
+                "--file",
+                str(pathlib.Path(__file__).parent.joinpath("describe-derivation.nix")),
+            ]
+            if arg is not None
+        ],
+        capture_output=True,
+        # write to file passed as argument (or stdout)
+        env=env,
+    )
+
+    description_str = description_process.stdout.decode()
+    if len(description_str) == 0:
+        logger.warning("Empty evaluation: %s", attribute_path)
+        logger.error(description_process.stdout)
+        return True
+
+    # add its inputs to the queue if they have not been processed
+    try:
+        description: model.Package = model.Package.parse_raw(description_str)
+    except Exception as e:
+        logger.warning(
+            "Failed to parse to model: attribute_path=%s, str=",
+            attribute_path,
+            description_str,
+        )
+        logger.error(description_process.stdout)
+        raise e from e
+
+    if description.output_path is None:
+        logger.warning("Ignore empty outputPath: %s", attribute_path)
+        return True
+
+    # write to output pipe
+    with pipe_out_lock:
+        pipe_out.write(description.json(by_alias=False))
+        pipe_out.write("\n")
+
+    visited_output_paths.add(description.output_path)
+    for build_input in description.build_inputs:
+        output_path = build_input.output_path
+        attribute_path = build_input.attribute_path
+        if output_path is not None and output_path not in queued_output_paths:
+            queue.put(attribute_path)
+            queued_output_paths.add(output_path)
+
+    # return success
+    return True
+
+
+def try_or_none(f: Callable, e_type: Type[Exception]):
+    try:
+        return f()
+    except e_type:
+        return None
+
+
+def queue_processor(
```

Comment:
I think it would be nice to mention that this function runs in parallel with `finder_output_reader` above. Or to write some class-level documentation mentioning that there are functions running in different threads in this file and give a high-level overview.

I know that this it normally outside the scope of this review, but actually the review requires looking here :slightly_smiling_face:

Reply:
Good point, I've opted for class-level documentation.